### PR TITLE
chicken: fix cross-compilation

### PR DESCRIPTION
**Pull request by siraben**
###### Motivation for this change
- [x] `nix -build -A pkgs.chicken` (build == host == target == x86_64-apple-darwin)
- [ ] `nix-build --argstr system 'x86_64-linux' -A pkgs.chicken` (build == host == target)
- [x] `nix-build --argstr system 'x86_64-linux' -A pkgs.pkgsCross.remarkable1.chicken` (build != host == target)
- [ ] `nix-build --argstr system 'x86_64-linux' -A pkgs.pkgsCross.remarkable1.buildPackages.chicken` (build == host != target)
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by siraben** [comment]
This is why `nix-build --argstr system 'x86_64-linux' -A pkgs.chicken` fails, this is the `defaults.make` file in the source.

```make
ifdef HOSTSYSTEM
C_COMPILER ?= $(HOSTSYSTEM)-gcc
CXX_COMPILER ?= $(HOSTSYSTEM)-g++
LIBRARIAN ?= $(HOSTSYSTEM)-ar # problematic line
else
C_COMPILER ?= gcc
CXX_COMPILER ?= g++
LIBRARIAN ?= ar
endif
```

The cross-compiled gcc is named `x86_64-unknown-linux-gnu-gcc-ar`, but the script looks for `x86_64-unknown-linux-gnu-ar` which fails.

File: pkgs/development/compilers/chicken/5/chicken.nix [hunk 1]
```diff
@@ -26,8 +24,16 @@ stdenv.mkDerivation {
   # -fno-strict-overflow is not a supported argument in clang on darwin
   hardeningDisable = lib.optionals stdenv.isDarwin ["strictoverflow"];
 
+  enableParallelBuilding = true;
+
+  patchPhase = ''
+    substituteInPlace defaults.make --replace '$(HOSTSYSTEM)-ar' '$(HOSTSYSTEM)-gcc-ar'
```
Comment (veprbl) [inline pkgs/development/compilers/chicken/5/chicken.nix:30]: Would it not work to just set the `LIBRARIAN` in `makeFlags`?

File: pkgs/development/compilers/chicken/5/chicken.nix [hunk 2]
```diff
@@ -26,8 +24,16 @@ stdenv.mkDerivation {
   # -fno-strict-overflow is not a supported argument in clang on darwin
   hardeningDisable = lib.optionals stdenv.isDarwin ["strictoverflow"];
 
+  enableParallelBuilding = true;
```
Comment (r-burns) [inline pkgs/development/compilers/chicken/5/chicken.nix:27]:
```suggestion
```
Fails immediately and consistently on my system, there's a race condition in creation of chicken-config.h

```
chicken.h:48:11: fatal error: chicken-config.h: No such file or directory
```

**Review by r-burns — CHANGES_REQUESTED** [review]
Getting this error when building (without parallelism) on non-nixos x86_64
```
x86_64-unknown-linux-gnu-gcc  -shared -Wl,-soname,libchicken.so.11 \
  -o libchicken.so library.o eval.o read-syntax.o repl.o data-structures.o pathname.o port.o file.o extras.o lolevel.o tcp.o srfi-4.o continuation.o posixunix.o internal.o irregex.o scheduler.o debugger-client.o profiler.o stub.o expand.o modules.o chicken-syntax.o chicken-ffi-syntax.o build-version.o runtime.o -lm -ldl
/nix/store/p3kn26g5nhmx6spn37ar3mn0xjbzks3g-binutils-2.31.1/bin/ld: cannot find crti.o: No such file or directory
collect2: error: ld returned 1 exit status
make: *** [rules.make:190: libchicken.so] Error 1
```

**Comment by siraben** [comment]
Using the compiler on the target doesn't seem to work,

```
reMarkable: ~/ /nix/store/j9s0lgiszr2sy8x270hdhgjqsab432ni-chicken-5.2.0-armv7l-unknown-linux-gnueabihf/bin/csc hello.scm                                                            
sh: armv7l-unknown-linux-gnueabihf-armv7l-unknown-linux-gnueabihf-gcc: command not found

Error: shell command terminated with non-zero exit status 32512: 'armv7l-unknown-linux-gnueabihf-armv7l-unknown-linux-gnueabihf-gcc' 'hello.c' -o 'hello.o' -c  -fno-strict-aliasing -fwrapv -DHAVE_CHICKEN_CONFIG_H -DC_ENABLE_PTABLES -Os -fomit-frame-pointer -I/nix/store/j9s0lgiszr2sy8x270hdhgjqsab432ni-chicken-5.2.0-armv7l-unknown-linux-gnueabihf/include/chicken
```

**Comment by siraben** [comment]
@r-burns I can't replicate that. Could you try again?

**Comment by siraben** [comment]
Relevant section of the docs on cross-compilation

```
==== Building the target libraries

First you need a version of the runtime system ({{libchicken}}),
compiled for the target system. Obtain and unpack a tarball of the
CHICKEN sources, or check out the code from the official code
repository, then build the libraries and necessary development files:

 make ARCH= \
     PREFIX=/usr \
     PLATFORM=linux \
     HOSTSYSTEM=arm-none-linux-gnueabi \
     DESTDIR=$HOME/target \
     TARGET_FEATURES="-no-feature x86 -feature arm" \
     install

```

**Comment by Ericson2314** [comment]
@siraben 
I think we should make a patch that's something like
```makefile
ifneq TARGETSYSTEM HOSTSYSTEM
TOOL_PREFIX ?= TARGETSYSTEM
else
TOOL_PREFIX ?=
endif

C_COMPILER ?= $(TOOL_PREFIX)gcc
CXX_COMPILER ?= $(TOOL_PREFIX)g++
LIBRARIAN ?= $(TOOL_PREFIX)ar # problematic line
```
But even that I suspect will not be enough, as the will need to distinguish between the tools used at runtime, and the tools used at build time, if, for example, they want to build libchicken in one go.

How slow is this build? A related good thing we can do is always build `libchicken` separately. This means moving bootstrapping logic from their build system to ours, which is good.

**Comment by siraben** [comment]
Hey @Mindavi, if you're free, could you also take a look at cross-compilation of CHICKEN as well? Let me know if you have any questions.

**Comment by Mindavi** [comment]
I'm running into this issue:

https://github.com/NixOS/nixpkgs/issues/116207 / https://github.com/NixOS/nixpkgs/issues/116504.

(oh, makeWrapper should be moved to nativeBuildInputs, good to know :+1: )

Haven't looked into this much further yet, that's for another day :).

**Comment by Artturin** [comment]
#145194

**Comment by nagy** [comment]
I have implemented this in #245420 . I am going to proclaim, that everything here has been done and close this PR.

**Pull request closed without merging**